extensions/stats/config.h

```diff
@@ -0,0 +1,384 @@
+
+struct Node {
+  common::NodeInfo node_info;
+  // key computed from the
+  std::string key;
+
+  Node(common::NodeInfo nodeInfo) {
+    auto labels = nodeInfo.labels();
+    absl::StrAppend(&key, nodeInfo.workload_name(), Sep, nodeInfo.namespace_(),
+                    Sep, labels["app"], Sep, labels["version"]);
+    node_info = nodeInfo;
+  }
+
+  Node() = delete;
+};
+
+using NodeSharedPtr = std::shared_ptr<Node>;
+
+class NodeInfoCache {
+ public:
+  // Fetches and caches Peer information by peerId
+  // TODO Remove this when it is cheap to directly get things from StreamInfo.
+  // At present this involves de-serializing to google.Protobuf.Struct and then
+  // another round trip to NodeInfo. This Should at most hold N entries.
+  const NodeSharedPtr getPeerById(StringView peerMetadataIdKey,
+                                  StringView peerMetadataKey);
+
+ private:
+  absl::flat_hash_map<std::string, NodeSharedPtr> cache_;
+};
+
+using valueExtractorFn = uint64_t (*)(const Common::RequestInfo& request_info);
+
+class SimpleStat {
+ public:
+  SimpleStat(uint32_t metric_id, valueExtractorFn value)
+      : metric_id_(metric_id), valueFn_(value){};
+
+  inline void record(const Common::RequestInfo& request_info) {
+    recordMetric(metric_id_, valueFn_(request_info));
+  };
+
+ private:
+  uint32_t metric_id_;
+  valueExtractorFn valueFn_;
+};
+
+using SimpleStatSharedPtr = std::shared_ptr<SimpleStat>;
+using MetricSharedPtr = std::shared_ptr<Metric>;
+
+#define SYMIFEMPTY(ex, sym) (ex).empty() ? (sym) : (ex)
+
+using mapperFn = std::function<std::string(
+    bool outbound, const common::NodeInfo& source, const common::NodeInfo& dest,
+    const Common::RequestInfo& requestInfo)>;
+
+struct Mapping {
+ public:
+  Mapping(std::string name, mapperFn mapper) : name_(name), mapper_(mapper){};
+
+  Mapping() = delete;
+
+  std::string name_;
+  mapperFn mapper_;
+};
+std::vector<Mapping> getStandardMappings();
+
+class Mappings {
+ public:
+  Mappings(std::vector<Mapping> mappings) : mappings_(mappings){};
+
+  Mappings() : Mappings(getStandardMappings()) {}
+
+  // metricTags converts mappings into ordered tags
+  std::vector<MetricTag> metricTags() {
+    std::vector<MetricTag> ret;
+
+    ret.reserve(mappings_.size());
+    for (const auto& mapping : mappings_) {
+      ret.push_back({mapping.name_, MetricTag::TagType::String});
+    }
+    return ret;
+  }
+
+  std::vector<std::string> eval(bool outbound, const common::NodeInfo& source,
+                                const common::NodeInfo& dest,
+                                const Common::RequestInfo& requestInfo) {
+    std::vector<std::string> vals;
+
+    vals.reserve(mappings_.size());
+    for (const auto& mapping : mappings_) {
+      vals.push_back(mapping.mapper_(outbound, source, dest, requestInfo));
+    }
+    return vals;
+  }
+
+ private:
+  std::vector<Mapping> mappings_;
+};
+
+#define MAPPING_SYM(key, expr, sym)                                      \
```

Comment:
I didn't see the lambda capture anything. Why there is lambda for each key? Can't you just create a lambda and use it every where?